### ISSUE TITLE
Server-Side Plugin System

### ISSUE THREAD

**Issue by KairuByte** [Phalcode/gamevault-backend#140]
**Is your feature request related to a problem? Please describe.**
There are many things that some may want to have happen on the server, that are separate from the intent of the app. Be it serving the downloads from a different server completely, serving games directly from a third party site (such as proxying from ichio) or many other situations.

**Describe the solution you'd like**
A plugin system would allow for these kinds of situations to be independently developed and supported.

**Comment by Alfagun74** [comment]
Technology i consider using: https://docs.nestjs.com/fundamentals/dynamic-modules
preparatory project restructuration: https://github.com/Phalcode/gamevault-backend/commit/2651afd2f5949ffaba403db134a50cbe1efeb20f

**Comment by Alfagun74** [comment]
I want the Plugin System to extend and change the basic logic of the application, but I'm struggling to find a way to do it.

**Comment by KairuByte** [comment]
The main method I know of to override basic logic is to literally load in libraries that hook into the code and run code before/after or even replace it. That said, I am not certain how to implement such a method into an application, as most methods I have dealt with were implemented by third parties to mod games and such (Harmony patching for instance.)

**Comment by Alfagun74** [comment]
Hey, guess what?! I've got some exciting news to share. I've discovered a fantastic method. Basically, we're allowing the server to execute any specified JavaScript code from defined sources, providing all the necessary information a plugin might need. And just for testing, I've created this amazing plugin:

[Chuck Norris Joke Fetcher Plugin](https://github.com/Phalcode/gamevault-backend-example-plugins/blob/master/plugins/gamevault-chuck-norris-joke-fetcher-plugin.js)

This plugin for example will run immediately upon loading.

Then, in the server configuration:
```env
PLUGIN_ENABLED=true
PLUGIN_SOURCES="https://raw.githubusercontent.com/Phalcode/gamevault-backend-example-plugins/master/plugins/gamevault-chuck-norris-joke-fetcher-plugin.js"
```



![Image](https://github.com/Phalcode/gamevault-backend/assets/22225051/24e28f9c-dd7f-4a31-8d9f-874b18b5989c)
![Server Image](https://github.com/Phalcode/gamevault-backend/assets/22225051/de7c439f-0a20-4c55-b347-332c2928430b)

Currently, the only context accessible to a plugin is the gamevault logger and the node fetch api to make HTTP requests. I still need to specify what other information a plugin should be able to access, such as the BoxArtService to retrieve box arts from your paid APIs.

I also need to establish a proper plugin schema so that machines can interpret them. Perhaps, we might even feature them on a self-updating marketplace in the store.

I'm really excited to see where this leads us as a community, as it could be a major fucking deal!
